cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java

```diff
@@ -0,0 +1,133 @@
+                                 ApplicationLifecycleService applicationLifecycleService,
+                                 NamespaceQueryAdmin namespaceQueryAdmin) {
+    this.preferencesService = preferencesService;
+    this.applicationLifecycleService = applicationLifecycleService;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+  }
+
+  @Path("/preferences")
+  @GET
+  public void getInstancePreferences(HttpRequest request, HttpResponder responder) {
+    PreferencesDetail detail = preferencesService.getPreferences();
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(detail, PreferencesDetail.class));
+  }
+
+  @Path("/namespaces/{namespace-id}/preferences")
+  @GET
+  public void getNamespacePreferences(HttpRequest request, HttpResponder responder,
+                                      @PathParam("namespace-id") String namespace, @QueryParam("resolved") boolean resolved) throws Exception {
+    NamespaceId namespaceId = new NamespaceId(namespace);
+    if (!namespaceQueryAdmin.exists(namespaceId)) {
+      throw new NamespaceNotFoundException(namespaceId);
+    }
+    PreferencesDetail detail;
+    if (resolved) {
+      detail = preferencesService.getResolvedPreferences(namespaceId);
+    } else {
+      detail = preferencesService.getPreferences(namespaceId);
+    }
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(detail, PreferencesDetail.class));
+  }
+
+  @Path("/namespaces/{namespace-id}/apps/{application-id}/preferences")
+  @GET
+  public void getApplicationPreferences(HttpRequest request, HttpResponder responder,
+                                        @PathParam("namespace-id") String namespace, @PathParam("application-id") String appId,
+                                        @QueryParam("resolved") boolean resolved) throws Exception {
+    ApplicationId applicationId = new ApplicationId(namespace, appId);
+    applicationLifecycleService.getAppDetail(applicationId);
+    PreferencesDetail detail;
+    if (resolved) {
+      detail = preferencesService.getResolvedPreferences(applicationId);
+    } else {
+      detail = preferencesService.getPreferences(applicationId);
+    }
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(detail, PreferencesDetail.class));
+  }
+
+  @Path("/namespaces/{namespace-id}/apps/{application-id}/{program-type}/{program-id}/preferences")
+  @GET
+  public void getProgramPreferences(HttpRequest request, HttpResponder responder,
+                                    @PathParam("namespace-id") String namespace, @PathParam("application-id") String appId,
+                                    @PathParam("program-type") String programType, @PathParam("program-id") String programId,
+                                    @QueryParam("resolved") boolean resolved) throws Exception {
+    ProgramId program = new ProgramId(namespace, appId, getProgramType(programType), programId);
+    applicationLifecycleService.getAppDetail(program.getParent());
```

Comment:
this should be returning a 404 if the app exists but the program does not

Reply:
this can be done with Store.ensureProgramExists()

Reply:
Done